### PR TITLE
Add VRAM profiling

### PR DESCRIPTION
**Pull request by drake7707**
Seeing how VRAM is affected by nodes is really useful for comparing between smaller and larger models

### CODE REVIEW

**Comment by drake7707** [comment]
image

File: nodes/execution_time.py [hunk 1]
```diff
@@ -26,16 +26,21 @@ def handle_execute(class_type, last_node_id, prompt_id, server, unique_id):
...
         return
     start_time = CURRENT_START_EXECUTION_DATA['nodes_start_perf_time'].get(unique_id)
+    start_vram = CURRENT_START_EXECUTION_DATA['nodes_start_vram_time'].get(unique_id)
     if start_time:
         end_time = time.perf_counter()
+        end_vram = model_management.get_free_memory()
         execution_time = end_time - start_time
+        # use abs because it can be negative, that means the model_manager already loaded the model when we started profiling and cleared the model at the end
+        # it's not super accurate for anything small but for large models it's really handy to know what the footprint is
+        vram_used = abs(end_vram - start_vram)
```
Comment (ty0x2333) [inline nodes/execution_time.py:36]: It is indeed possible that `end_vram` could be lower than `start_vram`, so using `max(0, end_vram - start_vram)` would be more accurate. However, this is not the main issue.

To truly calculate the VRAM usage, it should be the peak VRAM during the Node execution minus the VRAM at the start.

**Review by ty0x2333 — CHANGES_REQUESTED** [review]
Thank you for your contribution. I believe this will be a great feature.

However, there are a few minor issues in this PR. I’ve added comments in the PR.

@drake7707

File: nodes/execution_time.py (hunk 1 above)
Comment (drake7707) [reply nodes/execution_time.py:42]: >  To truly calculate the VRAM usage, it should be the peak VRAM during the Node execution minus the VRAM at the start.

I have no idea how to go about that unfortunately.  In a lot of cases the moment `nodes_start_vram_time` is set the node is already executing and the allocation has already been done, and the memory was already freed upon getting the end_vram so I almost always ended up with negative usage.

I suppose I could collect the minimum vram and maximum vram during the node execution and take that as usage, but I don't know if there is a callback from the server periodically or if I have to spawn a separate thread to do the memory polling. I wouldn't even know how to best do that in python (I'm not a python developer)

Comment (ty0x2333) [reply nodes/execution_time.py:42]: Sorry for noticing this late. I’ll first merge it into the other branch. I’ll make modifications later and release a new version afterwards.

**Pull request merged**